crates/driver/src/logic/competition/solution/approvals.rs

```diff
@@ -0,0 +1,29 @@
+use {crate::logic::eth, primitive_types::U256, std::collections::HashMap};
+
+/// A set of ERC20 approvals required by the [`super::Solution`]. This type
+/// guarantees that there is only one approval per [`eth::Spender`] and that the
+/// approvals are sorted deterministically.
+#[derive(Debug, Clone)]
+pub struct Approvals(Vec<eth::Approval>);
+
+impl Approvals {
+    /// Normalize the approvals such that there is only one approval per
+    /// [`eth::Spender`] and they are ordered deterministically.
+    pub fn normalize(approvals: impl Iterator<Item = eth::Approval>) -> Self {
+        let mut normalized = HashMap::new();
+        for approval in approvals {
+            let amount = normalized.entry(approval.spender).or_insert(U256::zero());
+            *amount = amount.checked_add(approval.amount).unwrap();
```

Comment:
The thing that tripped me up here is that `Approval` is an "approval request" indicating that "we need at least X amount of tokens approved for spender Y" and not the approval that should be set with an approval call.

Essentially, this turns into a:
```rust
if get_allowance(approval.spender) < approval.amount {
    Some(build_approval_interaction(approval.spender) // just spender, amount is U256::MAX
} else {
    None
}
```

And not:
```
build_approval_interaction(approval) // amount + spender
```

I would still not panic here because this condition could happen because of some **external** "invalid" data and not, for example, a logic bug. I would reserve the use of `panic`s for "this should never happen" and not "this can happen if an external system misbehaves".

Reply:
Maybe we should rename this to a `RequiredAllowance(s)` since it represents the minimum allowance that is needed and not the approvals that need to be made.

Reply:
> Why do the solvers return the approvals in /solve anyway? Isn't this something that the driver can determine based on orders? And furthermore, to make the lives of solvers easier, isn't this something that the driver should do?

The solvers propose `interactions`, which are these "block-box" calls to external contracts. Typically, when used for swapping some tokens against an on-chain AMM, the interaction will only work if there is the required approval given by the Settlement contract to some address specific to the on-chain AMM. In that regard, the approvals are not needed for users and orders (so, a settlement with perfect CoWs only including user orders will never need additional approvals).

Note that we include `inputs` and `outputs` with interactions, so in theory, we know the tokens and the amounts for the approvals that are needed. Unfortunately, we don't know the target. Some AMMs require approval directly on the contract that you interact for trading (Uniswap to their router contract for example), while others, require approvals to a completely different contract (ParaSwap swapper contract has a separate token approvals target that you need to set an allowance for).

> `RequiredAllowance` concept feels slightly unnatural

Interesting, maybe I've been in the space for so long, but this is a pretty "typical" concept to me. For example [0x includes this information in their swap response]() and [1Inch has a dedicated route to give you this information]().

One thing that we can do to make this more natural, is to attach an `approval_target: Address` to each `Interaction` or (since it needs has [input tokens and amounts](https://github.com/cowprotocol/services/blob/f87678d78cd431056e5b9ba2d1130ca41168b785/crates/shared/src/http_solver/model.rs#L150-L155)). This way it is closer to where it is actually needed (i.e. it is a pre-req for an interaction) and de-duplicates some of the data (token and amounts).

Reply:
Nice - this looks good.

We have a list of "Required" allowances from the solvers. These get normalized and summed up. Then these get compare to "Existing" allowances that have already been set, to generate the list of "Approval" transactions that need to be made.

I think this makes a bunch of sense and love the new naming! 💪